packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1451,6 +1461,25 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         }
     }
 
+    // Track how many times the container tries to reconnect with pending messages.
+    // This happens when the connection state is changed and we reset the counter
+    // when we are able to process a local op.
+    // If this counter reaches a max, it's a good indicator that the container
+    // is not making progress and it is stuck in a retry loop.
+    private shouldContinueReconnecting(): boolean {
+        if (this.maxConsecutiveReconnects < 0) {
+            // Feature disabled, we assume we are always making progress
+            return true;
+        }
+
+        return !this.pendingStateManager.hasPendingMessages()
+            || ++this.consecutiveReconnects < this.maxConsecutiveReconnects;
```

Comment:
Not sure how to do it, but I'd love to get telemetry that tells me something like - we can reduce defaultMaxConsecutiveReconnects  down from 15 to 5 without extra pain to the users (i.e., no false positives, and faster failure), 

Maybe issue telemetry event when we go through 1/2 of iterations? That way we can count number of sessions that hit it but did not close container - if it's close to zero, then we know we can half it and rinse & repeat?

Reply:
Can do!